qutip/solver/solver_base.py

```diff
@@ -95,7 +95,7 @@ def _restore_state(self, data, *, copy=True):
         """
         Retore the Qobj state from its data.
         """
-        if self._state_metadata['dims'] == self.rhs.dims[1]:
+        if self._state_metadata['dims'].as_list() == self.rhs._dims.as_list()[1]:
```

Comment:
Since both are Dimension object, the comparison should be able to work without the `as_list`. Did you try it?